action_server/src/action_server/actions/tell_name_of_person.py

```diff
@@ -0,0 +1,127 @@
+        self._preempt_requested = False
+
+        common_knowledge = load_knowledge('common')
+
+        self._grammar = ""
+
+        for name in common_knowledge.names:
+            self._grammar += "\nNAME['%s'] -> %s" % (name, name)
+
+    class Semantics:
+        def __init__(self):
+            self.location = None
+
+    @staticmethod
+    def _parse_semantics(semantics_dict):
+        semantics = TellNameOfPerson.Semantics()
+
+        if 'location' in semantics_dict:
+            semantics.location = resolve_entity_description(semantics_dict['location'])
+
+        return semantics
+
+    class Context:
+        def __init__(self):
+            self.object = None
+
+    @staticmethod
+    def _parse_context(context_dict):
+        context = TellNameOfPerson.Context()
+
+        if 'object' in context_dict:
+            context.object = resolve_entity_description(context_dict['object'])
+
+        return context
+
+    def _configure(self, robot, config):
+        self._robot = robot
+
+        semantics = TellNameOfPerson._parse_semantics(config.semantics)
+        context = TellNameOfPerson._parse_context(config.context)
+
+        # If a person is specified in the task description, we need to go and find that person first
+        if semantics.location and not context.object:
+            self._config_result.required_context = {
+                'action': 'find',
+                'object': {'type': 'person'},
+                'source-location': config.semantics['location'],
+            }
+            return
+
+        self._config_result.succeeded = True
+
+    def _start(self):
+        self._robot.head.look_at_standing_person()
+        self._robot.head.wait_for_motion_done()
+
+        self._robot.speech.speak("Hey sweetie what is your name?")
+
+        tries = 0
+        while tries < 3 and not self._preempt_requested:
+            try:
+                res = self._robot.hmi.query(description="",
+                                            grammar=self._grammar,
+                                            target='NAME')
+            except hmi.TimeoutException:
+                self._robot.speech.speak("My ears are not working properly, sorry!")
+                self._execute_result.message = "I was unable to hear anything when listening for your name"
+                tries += 1
+                continue
+
+            if res.semantics:
+                self._robot.speech.speak("Hi {name}! Bye {name}".format(name=res.sentence))
+                self._execute_result.message = "The person's name was {}".format(res.sentence)
+                self._execute_result.succeeded = True
+                return
+            else:
+                if tries < 2:
+                    self._robot.speech.speak("Sorry, I did not understand you, try again.")
+                else:
+                    self._robot.speech.speak("Sorry, I was unable to understand you again. I'll just name you Maple. ")
```

Comment:
Maple seems awfully Canada specific, maybe make it more general?

Reply:
The unisex name Max was implemented. (Consider this a tribute to you @MaxBaeten )

Reply:
This kind of behavior code should go to the smach states. Actions should never implement behavior, only logic to parse high level semantics from the command in it's context etc. Like we talked about the other week.

Reply:
Oops, never mind. The comment about doing this in another pr is still valid 😉